cpp/src/io/orc/stripe_data.cu

```diff
@@ -132,6 +132,146 @@ struct orcdec_state_s {
+                          ///< the first iteration of the top-level while-loop in
+                          ///< gpuDecodeOrcColumnData().
+    CAN_READ_FROM_CACHE,  ///< Run cache manager is ready for read. This is expected to happen in
+                          ///< the second iteration of the while-loop.
+  };
+
+ public:
+  /**
+   * @brief Initialize the run cache manager.
+   *
+   * @param[in] s ORC decoder state.
+   */
+  __device__ void initialize(orcdec_state_s* s)
+  {
+    _status          = (s->top.data.index.run_pos[CI_DATA2] > 0 and s->chunk.type_kind == TIMESTAMP)
+                         ? status::CAN_WRITE_TO_CACHE
+                         : status::DISABLED;
+    _reusable_length = 0;
+    _run_length      = 0;
+  }
+
+  /**
+   * @brief Set the reusable length object.
+   *
+   * @param[in] run_length The length of the first run (spanning two adjacent row groups) of the
+   * DATA stream.
+   * @param[in] max_length The maximum length allowed to be consumed. This limit is imposed
+   * by the decoder when processing the SECONDARY stream.
+   */
+  __device__ void set_reusable_length(uint32_t run_length, uint32_t max_length)
+  {
+    if (_status == status::CAN_WRITE_TO_CACHE) {
+      _run_length      = run_length;
+      _reusable_length = (_run_length > max_length) ? (_run_length - max_length) : 0;
+    }
+  }
+
+  /**
+   * @brief Adjust the maximum length allowed to be consumed when the length of the first run is
+   * greater than it.
+   *
+   * @param[in] max_length The maximum length allowed to be consumed for the DATA stream.
+   * @return A new maximum length.
+   */
+  [[nodiscard]] __device__ uint32_t adjust_max_length(uint32_t max_length)
+  {
+    auto new_max_length{max_length};
+    if (_status == status::CAN_READ_FROM_CACHE) { new_max_length -= _reusable_length; }
+    return new_max_length;
+  }
+
+  /**
+   * @brief Copy the excess data from the intermediate buffer for the DATA stream to the cache.
+   *
+   * @param[in] src Intermediate buffer for the DATA stream.
+   * @param[out] cache Local variable serving as the cache for the DATA stream.
+   */
+  __device__ void write_to_cache(int64_t* src, int64_t& cache)
+  {
+    if (_status != status::CAN_WRITE_TO_CACHE) { return; }
+
+    const auto tid = threadIdx.x;
+
+    // All threads in the block always take a uniform code path for the following branches.
+    // _reusable_length ranges between [0, 512].
+    if (_reusable_length > 0) {
+      const auto length_to_skip = _run_length - _reusable_length;
+      if (tid < _reusable_length) {
+        const auto src_idx = tid + length_to_skip;
+        cache              = src[src_idx];
+      }
+      // Block until all writes are done to safely change _status.
+      __syncthreads();
+      if (tid == 0) { _status = status::CAN_READ_FROM_CACHE; }
+    } else {
+      __syncthreads();
+      if (tid == 0) { _status = status::DISABLED; }
+    }
+
+    __syncthreads();
```

Comment:
__syncthreads() in conditionals are error prone. We can just sync outside as there are no inter-thread dependencies.

```suggestion
      if (tid == 0) { _status = status::CAN_READ_FROM_CACHE; }
    } else {
      if (tid == 0) { _status = status::DISABLED; }
    }

    __syncthreads();
```

Reply:
I moved the sync inside the if-else immediately above the if-else. The sync is still needed for correctness.